src/hangar/cli/cli.py

```diff
@@ -535,136 +536,172 @@ def server(overwrite, ip, port, timeout):
+@click.argument('startpoint', nargs=1, default=None, required=False)
+@click.option('-o', '--out', 'outdir', required=False, default=os.getcwd(), help="Directory to export data")
+@click.option('-s', '--sample', default=None, help='Sample name to export. You can optionally specify the type'
+                                                   ' of sample name after the colon (ex. `54:str` or `54:int`)')
+@click.option('-f', '--format', 'format_', required=False, help='File format of output file')
 @click.option('--plugin', required=False, help='override auto-inferred plugin')
 @pass_repo
-def export_data(repo: Repository, startpoint, arrayset, out, sample, format_, plugin):
-    """export ARRAYSET sample data as it existed a STARTPOINT to some format and path.
+@click.pass_context
+def export_data(ctx, repo: Repository, arrayset, outdir, startpoint, sample, format_, plugin):
+    """Export ARRAYSET sample data as it existed a STARTPOINT to some format and path.
+
+    Specifying which sample to be exported is possible by using the switch
+    `--sample` (without this, all the samples in the given arrayset will be
+    exported). Since hangar supports both int and str datatype for the sample
+    name, specifying that while mentioning the sample name might be necessary
+    at times. It is possible to do that by separating the name and type by a
+    colon.
+
+    Example:
+
+       1. if the sample name is string of numeric 10 - ``10:str`` or ``10``
+
+       2. if the sample name is ``sample1`` - ``sample1:str1`` or ``sample1``
+
+       3. if the sample name is an int, let say 10 - ``10:int``
     """
+    # TODO: ignore warning through env variable
     from hangar.records.commiting import expand_short_commit_digest
-    from hangar.records.heads import get_branch_head_commit
-    from hangar.cli.io import imsave
+    from hangar.records.heads import get_branch_head_commit, get_staging_branch_head
+    from hangar import external
+
+    kwargs = parse_custom_arguments(ctx.args)
 
     if startpoint in repo.list_branches():
         base_commit = get_branch_head_commit(repo._env.branchenv, startpoint)
-    else:
+    elif startpoint:
         base_commit = expand_short_commit_digest(repo._env.refenv, startpoint)
+    else:
+        branch_name = get_staging_branch_head(repo._env.branchenv)
+        base_commit = get_branch_head_commit(repo._env.branchenv, branch_name)
+    co = repo.checkout(commit=base_commit)
 
     try:
-        co = repo.checkout(write=False, commit=base_commit)
-        arrayset = co.arraysets[arrayset]
+        aset = co.arraysets.get(arrayset)
         if sample:
+            sample, stype = sample.split(':') if ':' in sample else (sample, '')
+            sample = int(sample) if stype == 'int' else str(sample)
             sampleNames = [sample]
         else:
-            sampleNames = list(arrayset.keys())
+            sampleNames = list(aset.keys())
```

Comment:
@hhsecond This is some seriously buggy code. 

- If i specify a sample `1:str` and `1:int`, when converted to a file name, they would be the same. 
 - same thing for if I don't specify file names, and there are mixed `str`, `int` keys, they would just end up overwriting eachother. 

Don't know what you'd like to do here, but this is bad that the tests aren't thorough enough to catch this.

I'll let you decide on what the fix is and implement it (along with the tests to prove it)

Reply:
The existing logic had changed here. Now instead of our CLI creating the file name, we let the plugin create the file names (this is because we are allowing writing to the same file in case of CSV, JSON etc). So I would assume plugin creators should handle such cases. If you think it makes sense, we could pass the type also along with sample name for the plugin to handle such cases but I would refrain from making another argument for the plugin function since there are workarounds. For answering your concerns
- I see the problem here but I would say, that does not seem like a frequent use-case (having sample name with both int and str of the same instance) and we could ignore it for now since I can't think of a sensible fix?
- The plugin could handle it
- So essentially, I skipped writing tests for it since it's a plugin thing

Reply:
> The existing logic had changed here. Now instead of our CLI creating the file name, we let the plugin create the file names (this is because we are allowing writing to the same file in case of CSV, JSON etc). So I would assume plugin creators should handle such cases.

Ok. This makes sense to me. 

> * I see the problem here but I would say, that does not seem like a frequent use-case (having sample name with both int and str of the same instance) and we could ignore it for now since I can't think of a sensible fix and since the plugin could handle it?
> * The plugin could handle it
> * So essentially, I skipped writing tests for it since it's a plugin thing

It doesn't matter if it's a frequent use case or not, it's something that we support (for better or worse in this instance) and since we won't control plugin code quality, we have to assume people will use this in strange ways with bad or just quickly written code. Sometime in the future when a piece of data overwrites itself on disk, the user is going to come and yell at us about the problem instead of going to the plugin developer. And honestly, i wouldn't really blame them, since they are using the `hangar` cli command after all. It's a minor thing now, and a lot of work to fix (clearly), but at the end of the day _every line of code has some degree of risk associated with it_, as library maintainers it's our job to minimize risk for our users. That's why they would use this over some home-grown solution at all.

As a side note, if you see a case where tests wouldn't pass because we don't have a piece of logic in place, it's better to write them and use `@pytest.mark.xfail(reason='foo', strict=True)` so we don't forget about smaller things like this it in the future. 

> So I would assume plugin creators should handle such cases.
> If you think it makes sense, we could pass the type also along with sample name for the plugin to handle such cases but I would refrain from making another argument for the plugin function since there are workarounds.

I agree this is a case where plugins creators handle this issue. It's just a matter of making a decision on how we should present this information to them. Passing a type along with the name isn't really any different from just passing a name as an `int` or `str`. If it's ignored on the plugin side, formatting the file name will just transparently cast and `int -> str`. 

We can eliminate this issue almost entirely is we do something like this:

1. we keep the same user facing selection API as discussed above 

```
    int:123   int
    str:123   str
    234       str
    foo1      str
    foo       str
    str:bar   str
```

But instead of passing a something we call `name` which would be either type `int` or `str`, every single sample name the plugin receives would be a string with the first 4 chars being `"type" + ":"`.

 It is up to that plugin to explicitly deal with how they think they should handle the names (in a way which is most convenient for that domain's users). If we just send them data like that, we are kindof forcing the plugin developers to deal with that decision and not get lazy. Even if they do, the user would have a file name with the type explicitly declared in it. 

```
    foo     -> 'str:foo'
    bar     -> 'str:bar'
    int:123 -> 'int:123'
    str:123 -> 'str:123'
    str:baz -> 'str:baz'
```

We document this behavior, and let the plugin authors deal with stripping the characters and formatting the name (or not, and it's OK for the user at the end of the day). 

Does this work?

Reply:
Thanks Rick, This is perfect